Flaky Test - TestP2PWithSingleDocumentSingleUpdateDoesNotSyncToNonPeerNode

**Describe the problem**
Another Flaky Test
```
...
✓  tests/integration/query/one_to_one_to_one (6.227s)
✖  tests/integration/net/state/simple/peer (27.161s) (-test.shuffle 1681310440125240211)
✓  tests/integration/query/latest_commits (13.868s)
...

=== Failed
=== FAIL: tests/integration/net/state/simple/peer TestP2PWithSingleDocumentSingleUpdateDoesNotSyncToNonPeerNode (unknown)
2023-04-12T10:40:54.985-0400, INFO, defra.tests.integration, , {"Database": "badger-in-memory"}
2023-04-12T10:40:55.165-0400, INFO, defra.tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-04-12T10:40:55.402-0400, INFO, defra.node, Created LibP2P host, {"PeerId": "12D3KooWBfE8GkNA8NeCX2uhfv6HSPnTutDcxTvFhPf25Vu9pVVQ", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-04-12T10:40:55.474-0400, INFO, defra.net, Starting internal broadcaster for pubsub network
2023-04-12T10:40:55.698-0400, INFO, defra.tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-04-12T10:40:56.011-0400, INFO, defra.node, Created LibP2P host, {"PeerId": "12D3KooWRz1tb7dnfWB8xT4xsJHtucjvkQDdPSyuRfYtnZ4tG4WR", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-04-12T10:40:56.032-0400, INFO, defra.net, Starting internal broadcaster for pubsub network
2023-04-12T10:40:56.087-0400, INFO, defra.tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-04-12T10:40:56.195-0400, INFO, defra.node, Created LibP2P host, {"PeerId": "12D3KooWNW8p48WZC2vckqnvxBzpzypvSRedSgYcjq2yH8R87ZiJ", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-04-12T10:40:56.220-0400, INFO, defra.net, Starting internal broadcaster for pubsub network
2023-04-12T10:40:58.466-0400, INFO, defra.tests.integration, Parsing bootstrap peers, {"Peers": "/ip4/172.27.231.49/tcp/43349/p2p/12D3KooWRz1tb7dnfWB8xT4xsJHtucjvkQDdPSyuRfYtnZ4tG4WR"}
2023-04-12T10:40:58.489-0400, INFO, defra.tests.integration, Bootstrapping with peers, {"Addresses": [{"ID":"12D3KooWRz1tb7dnfWB8xT4xsJHtucjvkQDdPSyuRfYtnZ4tG4WR","Addrs":["/ip4/172.27.231.49/tcp/43349"]}]}
2023-04-12T10:40:58.967-0400, INFO, defra.node, Connected, {"Peer ID": "12D3KooWRz1tb7dnfWB8xT4xsJHtucjvkQDdPSyuRfYtnZ4tG4WR"}
2023-04-12T10:40:58.994-0400, INFO, defra.net, Received new pubsub event, {"SenderId": "12D3KooWBfE8GkNA8NeCX2uhfv6HSPnTutDcxTvFhPf25Vu9pVVQ", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-04-12T10:40:59.011-0400, INFO, defra.net, Received new pubsub event, {"SenderId": "12D3KooWRz1tb7dnfWB8xT4xsJHtucjvkQDdPSyuRfYtnZ4tG4WR", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}

DONE 1309 tests, 1 failure in 89.464s
make: *** [Makefile:146: test] Error 1

```

**To Reproduce**
```
make test
```

**Expected behavior**
to not fail

--- a/net/peer.go
+++ b/net/peer.go
@@ -125,10 +125,7 @@ func NewPeer(
 	if err != nil {
 		return nil, err
 	}
-	err = p.loadP2PCollections(p.ctx)
-	if err != nil {
-		return nil, err
-	}
+
 	p.setupBlockService()
 	p.setupDAGService()
 
@@ -258,7 +255,7 @@ func (p *Peer) RegisterNewDocument(
 	)
 
 	// register topic
-	if err := p.server.addPubSubTopic(dockey.String(), true); err != nil {
+	if err := p.server.addPubSubTopic(dockey.String(), !p.server.hasPubSubTopic(schemaID)); err != nil {
 		log.ErrorE(
 			p.ctx,
 			"Failed to create new pubsub topic",
@@ -607,20 +604,21 @@ func (p *Peer) loadReplicators(ctx context.Context) error {
 	return nil
 }
 
-func (p *Peer) loadP2PCollections(ctx context.Context) error {
+func (p *Peer) loadP2PCollections(ctx context.Context) (map[string]struct{}, error) {
 	collections, err := p.db.GetAllP2PCollections(ctx)
 	if err != nil && !errors.Is(err, ds.ErrNotFound) {
-		return err
+		return nil, err
 	}
-
+	colMap := make(map[string]struct{})
 	for _, col := range collections {
 		err := p.server.addPubSubTopic(col, true)
 		if err != nil {
-			return err
+			return nil, err
 		}
+		colMap[col] = struct{}{}
 	}
 
-	return nil
+	return colMap, nil
 }
 
 func (p *Peer) handleDocCreateLog(evt events.Update) error {
@@ -629,10 +627,16 @@ func (p *Peer) handleDocCreateLog(evt events.Update) error {
 		return errors.Wrap("failed to get DocKey from broadcast message", err)
 	}
 
+	// We need to register the document before pushing to the replicators if we want to
+	// ensure that we have subscribed to the topic.
+	err = p.RegisterNewDocument(p.ctx, dockey, evt.Cid, evt.Block, evt.SchemaID)
+	if err != nil {
+		return err
+	}
 	// push to each peer (replicator)
 	p.pushLogToReplicators(p.ctx, evt)
 
-	return p.RegisterNewDocument(p.ctx, dockey, evt.Cid, evt.Block, evt.SchemaID)
+	return nil
 }
 
 func (p *Peer) handleDocUpdateLog(evt events.Update) error {
@@ -804,6 +808,30 @@ func (p *Peer) AddP2PCollections(collections []string) error {
 		addedTopics = append(addedTopics, col)
 	}
 
+	// If adding the collection topics succeeds, we remove the collections' documents
+	// from the pubsub topics to avoid receiving duplicate events.
+	for _, col := range collections {
+		c, err := store.GetCollectionBySchemaID(p.ctx, col)
+		if err != nil {
+			return err
+		}
+		keyChan, err := c.GetAllDocKeys(p.ctx)
+		if err != nil {
+			return err
+		}
+		for key := range keyChan {
+			err := p.server.removePubSubTopic(key.Key.String())
+			if err != nil {
+				log.Info(
+					p.ctx,
+					"Failed to remove doc from pubsub topic",
+					logging.NewKV("DocKey", key.Key.String()),
+					logging.NewKV("Cause", err),
+				)
+			}
+		}
+	}
+
 	return txn.Commit(p.ctx)
 }
 
@@ -848,6 +876,30 @@ func (p *Peer) RemoveP2PCollections(collections []string) error {
 				}
 			}
 			return err
+		}
+	}
+
+	// If removing the collection topics succeeds, we add back the collections' documents
+	// to the pubsub topics.
+	for _, col := range collections {
+		c, err := store.GetCollectionBySchemaID(p.ctx, col)
+		if err != nil {
+			return err
+		}
+		keyChan, err := c.GetAllDocKeys(p.ctx)
+		if err != nil {
+			return err
+		}
+		for key := range keyChan {
+			err := p.server.addPubSubTopic(key.Key.String(), true)
+			if err != nil {
+				log.Info(
+					p.ctx,
+					"Failed to add doc to pubsub topic",
+					logging.NewKV("DocKey", key.Key.String()),
+					logging.NewKV("Cause", err),
+				)
+			}
 		}
 	}
 

--- a/net/peer.go
+++ b/net/peer.go
@@ -767,6 +767,8 @@ type EvtPubSub struct {
 //
 // It will error if any of the given collectionIDs are invalid, in such a case some of the
 // changes to the server may still be applied.
+//
+// WARNING: Calling this on collections with a large number of documents may take a long time to process.
 func (p *Peer) AddP2PCollections(collections []string) error {
 	txn, err := p.db.NewTxn(p.ctx, false)
 	if err != nil {
@@ -839,6 +841,8 @@ func (p *Peer) AddP2PCollections(collections []string) error {
 //
 // It will error if any of the given collectionIDs are invalid, in such a case some of the
 // changes to the server may still be applied.
+//
+// WARNING: Calling this on collections with a large number of documents may take a long time to process.
 func (p *Peer) RemoveP2PCollections(collections []string) error {
 	txn, err := p.db.NewTxn(p.ctx, false)
 	if err != nil {

--- a/net/server.go
+++ b/net/server.go
@@ -51,6 +51,11 @@ type server struct {
 
 	pubSubEmitter  event.Emitter
 	pushLogEmitter event.Emitter
+
+	// docQueue is used to track which documents are currently being processed.
+	// This is used to prevent multiple concurrent processing of the same document and
+	// limit unecessary transaction conflicts.
+	docQueue *docQueue
 }
 
 // pubsubTopic is a wrapper of rpc.Topic to be able to track if the topic has
@@ -68,6 +73,9 @@ func newServer(p *Peer, db client.DB, opts ...grpc.DialOption) (*server, error) 
 		conns:  make(map[libpeer.ID]*grpc.ClientConn),
 		topics: make(map[string]pubsubTopic),
 		db:     db,
+		docQueue: &docQueue{
+			docs: make(map[string]chan struct{}),
+		},
 	}
 
 	cred := insecure.NewCredentials()
@@ -78,20 +86,30 @@ func newServer(p *Peer, db client.DB, opts ...grpc.DialOption) (*server, error) 
 
 	s.opts = append(defaultOpts, opts...)
 	if s.peer.ps != nil {
+		colMap, err := p.loadP2PCollections(p.ctx)
+		if err != nil {
+			return nil, err
+		}
+
 		// Get all DocKeys across all collections in the DB
 		log.Debug(p.ctx, "Getting all existing DocKey...")
-		keyResults, err := s.listAllDocKeys()
+		cols, err := s.db.GetAllCollections(s.peer.ctx)
 		if err != nil {
-			return nil, errors.Wrap("failed to get DocKeys for pubsub topic registration", err)
+			return nil, err
 		}
 
 		i := 0
-		if keyResults != nil {
-			for key := range keyResults {
-				if key.Err != nil {
-					log.ErrorE(p.ctx, "Failed to get a key to register pubsub topic", key.Err)
-					continue
-				}
+		for _, col := range cols {
+			// If we subscribed to the collection, we skip subscribing to the collection's dockeys.
+			if _, ok := colMap[col.SchemaID()]; ok {
+				continue
+			}
+			keyChan, err := col.GetAllDocKeys(p.ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			for key := range keyChan {
 				log.Debug(
 					p.ctx,
 					"Registering existing DocKey pubsub topic",
@@ -140,6 +158,38 @@ func (s *server) GetLog(ctx context.Context, req *pb.GetLogRequest) (*pb.GetLogR
 	return nil, nil
 }
 
+type docQueue struct {
+	docs map[string]chan struct{}
+	mu   sync.Mutex
+}
+
+// add adds a docKey to the queue. If the docKey is already in the queue, it will
+// wait for the docKey to be removed from the queue. For every add call, done must
+// be called to remove the docKey from the queue. Otherwise, subsequent add calls will
+// block forever.
+func (dq *docQueue) add(docKey string) {
+	dq.mu.Lock()
+	done, ok := dq.docs[docKey]
+	if !ok {
+		dq.docs[docKey] = make(chan struct{})
+	}
+	dq.mu.Unlock()
+	if ok {
+		<-done
+		dq.add(docKey)
+	}
+}
+
+func (dq *docQueue) done(docKey string) {
+	dq.mu.Lock()
+	defer dq.mu.Unlock()
+	done, ok := dq.docs[docKey]
+	if ok {
+		delete(dq.docs, docKey)
+		close(done)
+	}
+}
+
 // PushLog receives a push log request
 func (s *server) PushLog(ctx context.Context, req *pb.PushLogRequest) (*pb.PushLogReply, error) {
 	pid, err := peerIDFromContext(ctx)
@@ -150,6 +200,26 @@ func (s *server) PushLog(ctx context.Context, req *pb.PushLogRequest) (*pb.PushL
 
 	// parse request object
 	cid := req.Body.Cid.Cid
+
+	s.docQueue.add(req.Body.DocKey.String())
+	defer func() {
+		s.docQueue.done(req.Body.DocKey.String())
+		if s.pushLogEmitter != nil {
+			byPeer, err := libpeer.Decode(req.Body.Creator)
+			if err != nil {
+				log.Info(ctx, "could not decode the peer id of the log creator", logging.NewKV("Error", err.Error()))
+			}
+			err = s.pushLogEmitter.Emit(EvtReceivedPushLog{
+				FromPeer: pid,
+				ByPeer:   byPeer,
+			})
+			if err != nil {
+				// logging instead of returning an error because the event bus should
+				// not break the PushLog execution.
+				log.Info(ctx, "could not emit push log event", logging.NewKV("Error", err.Error()))
+			}
+		}
+	}()
 
 	// make sure were not processing twice
 	if canVisit := s.peer.queuedChildren.Visit(cid); !canVisit {
@@ -235,24 +305,15 @@ func (s *server) PushLog(ctx context.Context, req *pb.PushLogRequest) (*pb.PushL
 			return &pb.PushLogReply{}, txnErr
 		}
 
-		if s.pushLogEmitter != nil {
-			byPeer, err := libpeer.Decode(req.Body.Creator)
+		// Once processed, subscribe to the dockey topic on the pubsub network unless we already
+		// suscribe to the collection.
+		if !s.hasPubSubTopic(col.SchemaID()) {
+			err = s.addPubSubTopic(docKey.DocKey, true)
 			if err != nil {
-				log.Info(ctx, "could not decode the peer id of the log creator", logging.NewKV("Error", err.Error()))
-			}
-			err = s.pushLogEmitter.Emit(EvtReceivedPushLog{
-				FromPeer: pid,
-				ByPeer:   byPeer,
-			})
-			if err != nil {
-				// logging instead of returning an error because the event bus should
-				// not break the PushLog execution.
-				log.Info(ctx, "could not emit push log event", logging.NewKV("Error", err.Error()))
+				return nil, err
 			}
 		}
-
-		// Once processed, subscribe to the dockey topic on the pubsub network.
-		return &pb.PushLogReply{}, s.addPubSubTopic(docKey.DocKey, true)
+		return &pb.PushLogReply{}, nil
 	}
 
 	return &pb.PushLogReply{}, client.NewErrMaxTxnRetries(txnErr)
@@ -298,6 +359,14 @@ func (s *server) addPubSubTopic(topic string, subscribe bool) error {
 		subscribed: subscribe,
 	}
 	return nil
+}
+
+// hasPubSubTopic checks if we are subscribed to a topic.
+func (s *server) hasPubSubTopic(topic string) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.topics[topic]
+	return ok
 }
 
 // removePubSubTopic unsubscribes to a topic
@@ -406,48 +475,6 @@ func (s *server) pubSubEventHandler(from libpeer.ID, topic string, msg []byte) {
 			log.Info(s.peer.ctx, "could not emit pubsub event", logging.NewKV("Error", err.Error()))
 		}
 	}
-}
-
-func (s *server) listAllDocKeys() (<-chan client.DocKeysResult, error) {
-	// get all collections
-	cols, err := s.db.GetAllCollections(s.peer.ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	if len(cols) == 0 {
-		return nil, nil
-	}
-
-	keyCh := make(chan client.DocKeysResult)
-
-	var wg sync.WaitGroup
-	wg.Add(1) // add an init blocker on close routine
-	go func() {
-		wg.Wait()
-		close(keyCh)
-	}()
-
-	for _, col := range cols {
-		resCh, err := col.GetAllDocKeys(s.peer.ctx)
-		if err != nil {
-			return nil, err
-		}
-
-		// run a goroutine for each channel we get from the GetAllDocKeys func for each
-		// collection. Pipe the results from res to keys, and handle potentially
-		// closed channel edge cases
-		wg.Add(1)
-		go func(colName string) {
-			for res := range resCh {
-				keyCh <- res
-			}
-			wg.Done()
-		}(col.Name())
-	}
-	wg.Done() // cleanup the init blocker on close routine
-
-	return keyCh, nil
 }
 
 // addr implements net.Addr and holds a libp2p peer ID.

--- a/tests/integration/p2p.go
+++ b/tests/integration/p2p.go
@@ -140,6 +140,9 @@ func connectPeers(
 	nodes []*node.Node,
 	addresses []string,
 ) chan struct{} {
+	// If we have some database actions prior to connecting the peers, we want to ensure that they had time to
+	// complete before we connect. Otherwise we might wrongly catch them in our wait function.
+	time.Sleep(100 * time.Millisecond)
 	sourceNode := nodes[cfg.SourceNodeID]
 	targetNode := nodes[cfg.TargetNodeID]
 	targetAddress := addresses[cfg.TargetNodeID]
@@ -239,7 +242,9 @@ func connectPeers(
 	}
 
 	nodeSynced := make(chan struct{})
-	go func() {
+	ready := make(chan struct{})
+	go func(ready chan struct{}) {
+		ready <- struct{}{}
 		for waitIndex := 0; waitIndex < len(sourceToTargetEvents); waitIndex++ {
 			for i := 0; i < targetToSourceEvents[waitIndex]; i++ {
 				err := sourceNode.WaitForPushLogByPeerEvent(targetNode.PeerID())
@@ -251,7 +256,9 @@ func connectPeers(
 			}
 			nodeSynced <- struct{}{}
 		}
-	}()
+	}(ready)
+	// Ensure that the wait routine is ready to receive events before we continue.
+	<-ready
 
 	return nodeSynced
 }
@@ -285,6 +292,9 @@ func configureReplicator(
 	nodes []*node.Node,
 	addresses []string,
 ) chan struct{} {
+	// If we have some database actions prior to configuring the replicator, we want to ensure that they had time to
+	// complete before the configuration. Otherwise we might wrongly catch them in our wait function.
+	time.Sleep(100 * time.Millisecond)
 	sourceNode := nodes[cfg.SourceNodeID]
 	targetNode := nodes[cfg.TargetNodeID]
 	targetAddress := addresses[cfg.TargetNodeID]
@@ -307,7 +317,7 @@ func configureReplicator(
 				docIDsSyncedToSource[currentdocID] = struct{}{}
 			}
 
-			if action.NodeID.HasValue() && action.NodeID.Value() == cfg.SourceNodeID {
+			if !action.NodeID.HasValue() || action.NodeID.Value() == cfg.SourceNodeID {
 				sourceToTargetEvents[waitIndex] += 1
 			}
 
@@ -341,7 +351,9 @@ func configureReplicator(
 	}
 
 	nodeSynced := make(chan struct{})
-	go func() {
+	ready := make(chan struct{})
+	go func(ready chan struct{}) {
+		ready <- struct{}{}
 		for waitIndex := 0; waitIndex < len(sourceToTargetEvents); waitIndex++ {
 			for i := 0; i < targetToSourceEvents[waitIndex]; i++ {
 				err := sourceNode.WaitForPushLogByPeerEvent(targetNode.PeerID())
@@ -353,7 +365,9 @@ func configureReplicator(
 			}
 			nodeSynced <- struct{}{}
 		}
-	}()
+	}(ready)
+	// Ensure that the wait routine is ready to receive events before we continue.
+	<-ready
 
 	return nodeSynced
 }

--- a/tests/integration/p2p.go
+++ b/tests/integration/p2p.go
@@ -317,6 +317,8 @@ func configureReplicator(
 				docIDsSyncedToSource[currentdocID] = struct{}{}
 			}
 
+			// A document created on the source or one that is created on all nodes will be sent to the target even
+			// it already has it. It will create a `received push log` event on the target which we need to wait for.
 			if !action.NodeID.HasValue() || action.NodeID.Value() == cfg.SourceNodeID {
 				sourceToTargetEvents[waitIndex] += 1
 			}

